beefy-primitives/src/commitment.rs

```diff
@@ -95,6 +99,136 @@ impl<TBlockNumber, TPayload, TSignature> SignedCommitment<TBlockNumber, TPayload
+	/// in the uncompressed GRANDPA validators' signatures.
+	signatures_from: BitField,
+	/// Length of the `Vec` containing uncompressed GRANDPA validators' signatures.
+	signatures_len: u32,
+	/// A `Vec` containing all `TSignature` values present in the uncompressed
+	/// GRANDPA validators' signatures.
+	signatures_compact: Vec<TSignature>,
+}
+
+impl<'a, TBlockNumber, TPayload, TSignature>
+	CompactSignedCommitment<&'a Commitment<TBlockNumber, TPayload>, &'a TSignature>
+where
+	TSignature: Encode,
+	TBlockNumber: Encode,
+	TPayload: Encode,
+{
+	/// Packs a `SignedCommitment` into the compressed `CompactSignedCommitment` format for efficient network transport.
+	fn pack(signed_commitment: &'a SignedCommitment<TBlockNumber, TPayload, TSignature>) -> Self {
+		let SignedCommitment { commitment, signatures } = signed_commitment;
+		let signatures_len = signatures.len() as u32;
+		let mut signatures_from: BitField = vec![];
+		let mut signatures_compact: Vec<&TSignature> = vec![];
+
+		for signature in signatures {
+			match signature {
+				Some(value) => signatures_compact.push(value),
+				None => (),
+			}
+		}
+
+		let bits: Vec<u8> = signatures.iter().map(|x| if x.is_some() { 1 } else { 0 }).collect();
+		let chunks = bits.chunks(CONTAINER_BIT_SIZE);
+		for chunk in chunks {
+			let mut iter = chunk.iter().copied();
+			let mut v = iter.next().unwrap() as u8;
+
+			for bit in iter {
+				v <<= 1;
+				v |= bit as u8;
+			}
+
+			signatures_from.push(v);
+		}
+
+		Self {
+			commitment,
+			signatures_from,
+			signatures_len,
+			signatures_compact,
+		}
+	}
+}
+
+impl<TBlockNumber, TPayload, TSignature> Encode for SignedCommitment<TBlockNumber, TPayload, TSignature>
+where
+	TSignature: Encode,
+	TBlockNumber: Encode,
+	TPayload: Encode,
+{
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+		let temp = CompactSignedCommitment::pack(self);
+		temp.using_encoded(f)
+	}
+}
+
+impl<TBlockNumber, TPayload, TSignature> CompactSignedCommitment<Commitment<TBlockNumber, TPayload>, TSignature>
+where
+	TBlockNumber: Decode,
+	TPayload: Decode,
+	TSignature: Decode,
+{
+	/// Unpacks a `CompactSignedCommitment` into the uncompressed `SignedCommitment` form.
+	fn unpack(
+		temporary_signatures: CompactSignedCommitment<Commitment<TBlockNumber, TPayload>, TSignature>,
+	) -> SignedCommitment<TBlockNumber, TPayload, TSignature> {
+		let CompactSignedCommitment {
+			commitment,
+			signatures_from,
+			mut signatures_len,
+			signatures_compact,
+		} = temporary_signatures;
+		let mut bits: Vec<u8> = vec![];
+		let last_block = signatures_len % CONTAINER_BIT_SIZE as u32;
+		for block in signatures_from {
+			let start_bit = if signatures_len > last_block {
+				0
+			} else {
+				CONTAINER_BIT_SIZE - last_block as usize
+			};
+
+			for bit in start_bit..CONTAINER_BIT_SIZE {
+				let bit_position = CONTAINER_BIT_SIZE - bit - 1;
+				bits.push(block >> bit_position & 1);
+				signatures_len -= 1;
+			}
```

Comment:
Actually, disregard my previous comment - given the way we process `bits` later on, I think it's safe to not do any conditionals for the last block - it will simply be filled with `0`.

So the way I see this loop is that we just read everything from `signatures_from`. The inner loop should simply go `for bit in 0..CONTAINER_BIT_SIZE` and then we just trim resulting `signatures` vector to match `signatures_len`. That should simplify the code a lot - wdyt?

Reply:
@de-sh Please let me know if you need any more clarification on this. I think this is definitely worth doing and will simplify the code a lot.

Reply:
Awesome! Now I understand quite a lot better, thank you for the patience 😃

Reply:
Try this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dd88ffe4bc09e1d826ea5a6d9c6d0b90

Reply:
Instead of doing a slice we can also store the bits in reverse order and truncate (but then we also need to store compact signatures in reverse order too).

Reply:
Just noticed what looks like an edge case: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9c8971e090acacd84518149533e23e04

Reply:
This seems like a potential solution, I'll push it, please review.